### DISCUSSION TITLE
trouble getting python binding in windows 10  with vcpkg build

### DISCUSSION THREAD

**Discussion by mexxmillion** [OpenImageIO/oiio#3461] [Q&A] [answered]
hi
i have been using oiio with python binding in MacOS using brew and conda environments.
Now, i have tried to build it for windows 10 using the vcpkg and it seems to have finished successfully.
but i cant get the python binding to work.

Installation using this:
vcpkg install openimagio[ffmpeg,libraw,opencolorio,pybind11]:x64-windows-static

i can see the **OpenImageIO.cp310-win_amd64.pyd** in my user folder
_C:\Users\User\git\vcpkg\packages\openimageio_x64-windows\lib\python3.10\site-packages_

I have set this in my PYTHONPATH env variable too.

so now if i start my conda env. using  this python
_"Python 3.10.5 | packaged by conda-forge | (main, Jun 14 2022, 06:57:19) [MSC v.1929 64 bit (AMD64)] on win32"_

>>> import OpenImageIO
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'OpenImageIO'

if i do
print(sys.path)
['', 'C:\\Users\\User\\git\\vcpkg\\packages\\openimageio_x64-windows\\lib\\python3.10', 'C:\\Users\\User\\anaconda3\\envs\\py310', 'C:\\Users\\User', 'C:\\Users\\User\\anaconda3\\envs\\py310\\python310.zip', 'C:\\Users\\User\\anaconda3\\envs\\py310\\DLLs', 'C:\\Users\\User\\anaconda3\\envs\\py310\\lib', 'C:\\Users\\User\\anaconda3\\envs\\py310\\lib\\site-packages', 'C:\\Users\\User\\anaconda3\\envs\\py310\\lib\\site-packages\\win32', 'C:\\Users\\User\\anaconda3\\envs\\py310\\lib\\site-packages\\win32\\lib', 'C:\\Users\\User\\anaconda3\\envs\\py310\\lib\\site-packages\\Pythonwin']
>>>

what do i need to make OpenImageIO python binding avail for me on windows?
also, I am not experienced with any build, how do i get it work build for specific python like 3.8 or 3.9 as it build 3.10 by default.

thanks

**Comment by lgritz** [comment]
I'm not a python on Windows expert, so I'm hoping somebody else can also chime in.

As far as not finding the module, my first guess would be to make sure your PYTHONPATH environment variable contains the directory where the module ends up getting installed?

**Comment by Daveblueband** [comment, marked as answer]
I can speak to this issue as I was having similar problems in the last year due to moving to python 3.8 (and recently 10). In Python 3.8, they changed how DLL resolution can safely be handled on Windows. I recommend reviewing this post which poses the generic problem of DLL loading and the one answer cites the Python 3.8 What’s New Docs as well as an example of setting it using the os.add_dll_directory() function. Basically modifying the PATH env var will no longer work in 3.8+ for DLL resolution (It of course still works in 3.7 and older). 

https://stackoverflow.com/questions/214852/python-module-dlls

Another thing I’ve done in my VCPKG build testing is that the PYD and DLLs are in separate build folders. If you build the CMake way, they are in the same directories so I’ve determined that VCPKG is being overly clean about how it’s builds are created and thus I can copy the PYD into the same directory of the DLLs (or better yet, extract the contents to a clean directory away from re-building folders).

 As to specific Python version builds against VCPKG, this is a little tricky. They recently added a ‘versions’ mechanism for allowing a user to essentially provide config rules as to what to use, but I’ve found it to be not always reliable as any dependencies that also require Python will have to have had snapshots of those python versions to source from (PyBind11, for example, I’ve not been able to build via VCPKG with python 3.7 due to a snapshot issue and thus overall I would be wary building OIIO with 3.7 via VCPKG). Personally, I recommend building with CMake OIIO with the python you absolutely need and worse case using VCPKG to get non-python dependent libraries to point to in your CMake. It gives the most flexibility in choosing your specific needs, but agree it is not the easiest to do. I can confirm I have built for Python 3.8, 3.9 and 3.10 this way before they were supported by default VCPKG builds.

https://devblogs.microsoft.com/cppblog/take-control-of-your-vcpkg-dependencies-with-versioning-support/

Hope this helps!

**Reply by mexxmillion** [reply]
Thank you so much @Daveblueband . I managed to make it work by copying the .pyd into "tools/openimagio" folder and setting that in my PYTHONPATH.
For my own personal test, I will use with 3.10 and figure out how to compile for 3.9 later.

@lgritz Thanks for making such a useful tool that i use almost everyday at work and home. 
hoping its getting easier for casual newbie like me to install on windows soon.

**Reply by stereorobin** [reply]
Thank you Dave for the thorough explanation. I managed to piece together that the PYD should be in the same folder as the DLLs from reading a lot of old posts online, wish I had come across your answer sooner.
Maybe it would be good to add this information under the "Building on Windows" section in the INSTALL.MD? From what I've seen I think a lot of people run into the same dead end when trying to follow the instructions and when asking for help the response is usually "make sure your pythonpath is set".

@mexxmillion Did you figure out how to compile for 3.9? That was going to be my next step now that I got it running on 3.10.
